### ISSUE TITLE
Error with the RTDB for when I attempt to use vdw disp corrections

### ISSUE THREAD

**Issue by viviennep** [nwchemgit/nwchem#692]
I consistently get the error: `bgj_get_scf_method: error reading rtdb  0` when attempting to use any of the vdW corrections via `disp vdw #` (every number causes the error). 
A google search only returns some forum posts from 2011 & 2012, without any solution (the issue didn't arise because of vdw corrections back then either). 

Because I was getting this issue on a larger system, I attempted to play with the settings on a smaller system to see if I could fix it. First, I made a run with a water molecule using the tpss03 x & c functionals with the `disp vdw 4` correction, and it ran without error! It definitely used the D3(BJ) correction, too; the scale factors were reported in the stdout. So, I then attempted to make a cluster of water molecules, I put nine of them near each other, and this caused the issue to come back. I started to remove water molecules one by one until the error stopped occurring, which happened when I only had 4 water molecules left. 

The input is barebones:
```
start vdw
echo

Title "vdW Test"

geometry  units angstrom noautoz noautosym
O     -0.798477   -3.639442   -3.619079
H     -0.103981   -3.594787   -4.265748
H     -1.458044   -4.260457   -3.905106
O     -0.732309   -1.756391   -3.336247
H     -1.218624   -2.217078   -2.662627
H     -0.753694   -2.259081   -4.142067
O     -0.199019   -1.965064   -0.890623
H     -1.119142   -2.193873   -0.831298
H      0.025011   -1.368506   -0.186046
O     -0.655166    1.883948   -3.022116
H     -1.501368    2.297319   -3.146881
H     -0.779630    0.969823   -2.795433
O     -2.655166    1.883948   -3.022116 # Comment this
H     -3.501368    2.297319   -3.146881 # water out and
H     -2.779630    0.969823   -2.795433 # it works
end

basis; * library 6-31g*; end

dft; print debug; xc xtpss03 ctpss03; disp vdw 4; end

task dft optimize ignore  
```
For reference, changing the xc functionals and/or changing the basis set has no effect, as I would expect for these corrections. 

I compiled this nwchem binary with a clone I made of this repo yesterday using gcc/gfortran 12.1 & openmpi 4.1.3 on a Linux cluster. This was run via the slurm scheduler, and using a single core vs multiple doesn't change anything. 
  
Every run I've described was done with a clean directory as well, with all .db and other output files removed prior to re-running.

Attached is the output associated with that run: [stdout.log](https://github.com/nwchemgit/nwchem/files/9998520/stdout.log)

I'll put the relevant error here:
```
 *** fock_2e: xc_active  F
 ------------------------------------------------------------------------
 bgj_get_scf_method: error reading rtdb                   0
 ------------------------------------------------------------------------
 ------------------------------------------------------------------------
  current input line :
    28: task dft optimize ignore
 ------------------------------------------------------------------------
 ------------------------------------------------------------------------
 An error occured in the Runtime Database
 ------------------------------------------------------------------------
 For more information see the NWChem manual at https://nwchemgit.github.io


 For further details see manual                                                                                                                                                         section:                                                                                                                                                                                
datum_fread: failed to read data 8
rtdb_seq_get: get failed for "bgj:scf_type" in ./vdw.db
Last System Error Message from Task 0:: Bad address
```
Also, here's the vdw.db rtdb file that gets created (I added the fake .log extension to it so that I could attach it to github easier): [vdw.db.log](https://github.com/nwchemgit/nwchem/files/9998530/vdw.db.log)
Looking at this with xxd, the relevant lines containing `bgj scf type` are these, where I've surrounded the words following the two "scf_type" instances with `*`'s:
```
000c20b0: 0000 0008 0000 0001 0000 0062 676a 3a73    ...........bgj:s
000c20c0: 6366 5f74 7970 65*00 02*00 0000 0000 0000  cf_type*..*.......
000c20d0: 1300 0000 2400 0000 0100 0000 2172 7464    ....$.......!rtd
000c20e0: 6221 6267 6a3a 7363 665f 7479 7065 *00f2*  b!bgj:scf_type*..*
```
I do this because [this](https://nwchemgit.github.io/Special_AWCforum/st/id302/bgj_get_scf_method__error_readin....html) forum thread recommended hacking a 2 into the rtdb where it might be a 3 if a hybrid functional gets used... Obviously that's not the case here because it looks like it's already set to 2.

Would love any help y'all!

**Comment by edoapra** [comment]
@viviennep Is this the output of a running using multiple processes or a single process?

**Comment by viviennep** [comment]
The attached log file was from a run with a single process.

**Comment by edoapra** [comment]
Do you what  linux distribution is used on the cluster you are using?
Could you post the output of 
gcc -v
gfortran -v

**Comment by edoapra** [comment]
If you are able to upload the full log of the compilation of NWChem, that would be very helpful.

**Comment by viviennep** [comment]
I don't believe that I have a full log of the compilation, sadly. The distro is Rocky Linux, and I used [these environment variables](https://github.com/nwchemgit/nwchem/files/9999025/env_script.txt) 


Here's gcc -v
```
Reading specs from /packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib/gcc/x86_64-pc-linux-gnu/12.1.0/specs
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/libexec/gcc/x86_64-pc-linux-gnu/12.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /tmp/software/spack-stage/spack-stage-gcc-12.1.0-hxmipkh42gkwphldmgf52vo37v6cq4q7/spack-src/configure --prefix=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm --with-pkgversion='Spack GCC' --with-bugurl=https://github.com/spack/spack/issues --disable-multilib --enable-languages=c,c++,fortran --disable-nls --disable-canonical-system-headers --with-system-zlib --with-zstd-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/include --with-zstd-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib --enable-bootstrap --with-mpfr-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/include --with-mpfr-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib --with-gmp-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/include --with-gmp-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib --with-mpc-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/include --with-mpc-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib --without-isl --with-stage1-ldflags='-Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib64 -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zlib-1.2.12-m72/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib' --with-boot-ldflags='-Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib64 -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zlib-1.2.12-m72/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib -static-libstdc++ -static-libgcc' --with-build-config=spack
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (Spack GCC)
```
and gfortran -v 
```
Reading specs from /packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib/gcc/x86_64-pc-linux-gnu/12.1.0/specs
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/libexec/gcc/x86_64-pc-linux-gnu/12.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /tmp/software/spack-stage/spack-stage-gcc-12.1.0-hxmipkh42gkwphldmgf52vo37v6cq4q7/spack-src/configure --prefix=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm --with-pkgversion='Spack GCC' --with-bugurl=https://github.com/spack/spack/issues --disable-multilib --enable-languages=c,c++,fortran --disable-nls --disable-canonical-system-headers --with-system-zlib --with-zstd-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/include --with-zstd-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib --enable-bootstrap --with-mpfr-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/include --with-mpfr-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib --with-gmp-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/include --with-gmp-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib --with-mpc-include=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/include --with-mpc-lib=/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib --without-isl --with-stage1-ldflags='-Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib64 -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zlib-1.2.12-m72/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib' --with-boot-ldflags='-Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gcc-12.1.0-hxm/lib64 -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/gmp-6.2.1-tun/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpc-1.2.1-lw7/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/mpfr-4.1.0-m7h/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zlib-1.2.12-m72/lib -Wl,-rpath,/packages/apps/spack/18/opt/spack/gcc-11.2.0/zstd-1.5.2-5qt/lib -static-libstdc++ -static-libgcc' --with-build-config=spack
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (Spack GCC)
```
---

**Comment by edoapra** [comment]
It sounds like gcc/gfortran are provided as spack modules, right?

As an alternative, do you have the option of running containers on your cluster (e.g. Docker, Singularity, Shifter, etc  ...)?

**Comment by viviennep** [comment]
I do not believe that I have that option sadly. 
I wouldn't mind recompiling but saving the logs this time if you think that would be helpful.
Alternatively, I can ask an admin to install with spack if that is a more guaranteed correct method.

**Comment by edoapra** [comment]
> I do not believe that I have that option sadly. Alternatively, I can ask an admin to install with spack if that is a more guaranteed correct method.

Please do not try that option.  My only "unproven" hypothesis is that there is something not working the software stack from spack you are using ... therefore using the NWChem spack build is not going to be any better

> As for the full log, I just recompiled and saved the log. I tested and the error persists. [Here](https://github.com/nwchemgit/nwchem/files/10000883/make.log) is the compile log file, and [here](https://github.com/nwchemgit/nwchem/files/10000885/stdout.log) is output of the above job with this new binary showing the same error.

Thanks for providing this.  
I have not spotted any clear problem yet.  
The only error (that should not cause the rtdb failure) on the log is  
```
make[2]: Entering directory '/home/rpelleti/compilations/nwchem/src/tce'
(standard_in) 1: syntax error
```
Could you please send the following
* output of the commands `make -v`  and `which make`
*  output of the command `cd $NWCHEM_TOP/src/tce; git diff GNUmakefile`
*  output of the command `cd $NWCHEM_TOP/src/tce; tail -20 GNUmakefile`

**Comment by edoapra** [comment]
Let's try to explore the I/O option.
Could you set the [permant_dir](https://nwchemgit.github.io/Permanent_Dir.html) and [scratch_dir](https://nwchemgit.github.io/Scratch_Dir.html) in the input file to a local file system on the compute node of your cluster?
/tmp or /dev/shm should work. This is the input file change I am alluding to
```
start vdw
echo
scratch_dir /dev/shm
permanent_dir /dev/shm
...
```

**Comment by viviennep** [comment]
Below I have the output to the make information which you want, but more importantly setting the scratch and permanent dirs to `/dev/shm` runs without error! Here is the output for the same job
[stdout_shm.log](https://github.com/nwchemgit/nwchem/files/10006618/stdout_shm.log)


As for make,
`which make` is `/usr/bin/make`, and `make -v` is
```
GNU Make 4.2.1
Built for x86_64-redhat-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

There is no difference from `git diff GNUmakefile`, and the last 20 lines are:
```
# Possible #defines
#
#              -DDEBUG_PRINT          /* Debug printing routines */
#              -DNOCOMMS              /* Turn off ALL communication calls */
#              -DBLOCK_TRANSF         /* Use block-transfer for comms calls */
#              -DOLD_AOBLOCK          /* Recover original AO blocking scheme */
#

HEADERS =
ifdef USE_64TO32
MAKEV_GE_4= $(shell echo $(MAKE_VERSION)\>=4.0 | bc )
ifeq ($(MAKEV_GE_4),1)
        override undefine CONVERT_ALL
endif
endif
include ../config/makefile.h
include ../config/makelib.h

ga_test: ga_test.o
        $(LINK.f) $< $(LIBS) -o $@
```

**Comment by edoapra** [comment]
> Below I have the output to the make information which you want, but more importantly setting the scratch and permanent dirs to `/dev/shm` runs without error! Here is the output for the same job [stdout_shm.log](https://github.com/nwchemgit/nwchem/files/10006618/stdout_shm.log)

That's good to know! Do you what kind of file system is used to mount the /home directory? Is it NFS?  

It's clear that the fread() operations used in NWChem's rtdb are failing when using /home

**Comment by viviennep** [comment]
This is interesting, runs which were failing were being done in the system's scratch directory which has a BeeGFS file system, my home directory is NFS and I just ran this job in my home directory and there was no error.

**Comment by edoapra** [comment]
Does this mean that your earlier failed jobs were using  BeeGFS both for scratch_ and permanent_dir ?

**Comment by viviennep** [comment]
Yes. I was using `.` as the location for both.

**Comment by edoapra** [comment]
Could you try to apply the attached patch and see if it fixes the BeeGFS issue?
To apply:
```
cd $NWCHEM_TOP
patch -p1 < rtdbgetflush.patch.txt
cd src/rtdb
make 
cd ..
make link
```
[rtdbgetflush.patch.txt](https://github.com/nwchemgit/nwchem/files/10007541/rtdbgetflush.patch.txt)

**Comment by viviennep** [comment]
This did not fix the issue.

**Comment by edoapra** [comment]
> This did not fix the issue.

Thank your very much for the feedback.

**Comment by edoapra** [comment]
The weird thing is that, while our local cluster has both a BeeGFS file system and a NFS file system, I am not able to reproduce your problem at all.  I do use with success the BeeGFS partition as permanent_dir.  
I am guessing that configuration of the BeeGFS must be responsible for your [Posix fread()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fread.html) failures.
I have checked the entries of the vdw.db file you have uploaded and they are as expected (therefore [fwrite()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fwrite.html) works correctly)

In summary, there is not much else I can do from the NWChem side to address this issue.
